replica/table.cc

```diff
@@ -974,17 +1002,34 @@ compaction_group& table::compaction_group_for_sstable(const sstables::shared_sst
     return _sg_manager->compaction_group_for_sstable(sst);
 }
 
-compaction_group_list& table::compaction_groups() const noexcept {
-    return _sg_manager->compaction_groups();
+future<> table::parallel_foreach_compaction_group(std::function<future<>(compaction_group&)> action) {
+    co_await _sg_manager->parallel_foreach_storage_group([&] (size_t, storage_group& sg) -> future<> {
+        co_await coroutine::parallel_for_each(sg.compaction_groups(), [&] (compaction_group* cg) -> future<> {
+            co_await action(*cg);
+        });
+    });
+}
+
+void table::for_each_compaction_group(std::function<void(compaction_group&)> action) {
+    _sg_manager->for_each_storage_group([&] (size_t, storage_group& sg) {
+        sg.for_each_compaction_group([&] (compaction_group& cg) {
+           action(cg);
+       });
+    });
 }
 
-future<> table::parallel_foreach_compaction_group(std::function<future<>(compaction_group&)> action) {
-    rwlock::holder shared_lock = co_await _sg_manager->get_rwlock().hold_read_lock();
-    co_await coroutine::parallel_for_each(compaction_groups(), [&] (compaction_group& cg) {
-        return action(cg);
+void table::for_each_const_compaction_group(std::function<void(const compaction_group&)> action) const {
```

Comment:
nit: do we really need a different function name?
isn't the const qualifier and the `action` type enough to differentiate between them?

Reply:
good question, the compiler should be able to tell the difference by looking at lambda param type. if you don't have any major comment, I am happy to send a patch as a followup.

Reply:
Seems to work. See https://godbolt.org/z/qM7P8xze9